### DISCUSSION TITLE
Version control for different deployment methods

### DISCUSSION THREAD

**Discussion by DuyTr** [gregsullivan/_tw#128] [General]
We want to build the build script `npm run production` into our deployment pipeline. GMD2024_VERSION never changes so cache busting files not working.

**Comment by gregsullivan** [comment]
Hello—Thanks for posting this!

`npm run production` builds production versions of all of the CSS and JavaScript files associated with your theme. _tw's built-in cache busting is part of `npm run bundle`, which creates a production-ready zip file for upload to a WordPress installation.

If you want cache busting, but you don't want to use (or can't use) `npm run bundle` and the resulting zip file, you can add your own versioning as part of your deployment pipeline. This is the relevant code:

https://github.com/gregsullivan/_tw/blob/fd500cc671105c5228e727b4b15236e38cbb0420/node_scripts/zip.js#L40-L45

During the process of creating the zip file, it does a search-and-replace for the constant definition containing the version and replaces its value with the current time (converted to seconds) in base 36. (This is based on how Drupal does—did?—cache busting on similar resources.)

I have seen people do cache busting by calculating an MD5 hash for the files each time they're enqueued, but that approach always made me uncomfortable. I'm sure it's not that computationally expensive a process, but I'd still prefer to do it as part of a build or deployment process.

Hope that helps! I'm going to convert this to a discussion, but please feel free to let me know if you have any other questions.

**Comment by tleifj** [comment]
Hi Greg,

If we're using `npm run prod` before we deploy styles and scripts to the server, it looks like the theme uses the theme version constant and appends that to the file URL when it's enqueued. So does that mean in order to do cache busting, we should be updating that theme constant and pushing up `functions.php` whenever we make style or script updates?

Thank you!
Travis

**Reply by gregsullivan** [reply]
Hi Travis—Thanks for your message, and very sorry for the delay here! I don't think GitHub sent me a notification about your message.

You're absolutely right: If you're just using `npm run prod`, and not using the `npm run bundle` approach that creates a zip file, you're responsible for your own cache busting via the version constant.

You can either manually update the constant before deployment, or you can automate updates in a similar way to the `./node_scripts/zip.js` script, as excerpted above. (In the latter case, the idea would be to update it as part of something like a GitHub Action, so that the new version constant isn't committed to GitHub, existing only in the deployed files for cache-busting purposes.)

Hope that helps!
